app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx

```diff
@@ -370,13 +365,143 @@ class SamplesHeatmapView extends React.Component {
       // We overwrite both this.state.sampleDetails and this.state.sampleIds to make sure the two are in sync.
       sampleIds,
       sampleDetails,
-      taxonIds,
-      taxonDetails,
-      data,
+      allTaxonIds,
+      allTaxonDetails,
+      allData,
       taxonFilterState,
     };
   }
 
+  filterTaxons() {
+    this.setTaxonThresholdFilterState();
+    let {
+      allTaxonIds,
+      allTaxonDetails,
+      taxonFilterState,
+      allData,
+    } = this.state;
+    let taxonDetails = {},
+      taxonIds = [],
+      filteredData = {};
+    if (this.props.allowedFeatures.includes("heatmap_filter_fe")) {
+      for (let taxonId of allTaxonIds) {
+        let taxon = allTaxonDetails[taxonId];
+        if (
+          !taxonIds.includes(taxonId) &&
+          this.taxonPassesSelectedFilters(taxon)
+        ) {
+          if (Object.values(taxonFilterState[taxon["index"]]).includes(true)) {
+            taxonDetails[taxon["id"]] = taxon;
+            taxonDetails[taxon["name"]] = taxon;
+            taxonIds.push(taxon["id"]);
+            this.props.metrics.forEach(metric => {
+              filteredData[metric.value] = filteredData[metric.value] || [];
+              filteredData[metric.value][filteredData[metric.value].length] =
+                allData[metric.value][taxon["index"]] || [];
+            });
+          }
+        }
+      }
+    } else {
+      taxonDetails = allTaxonDetails;
+      taxonIds = allTaxonIds;
+      filteredData = allData;
+    }
+
+    this.updateHistoryState();
+
+    this.setState({
+      taxonDetails: taxonDetails,
+      taxonIds: taxonIds,
+      loading: false,
+      data: filteredData,
+    });
+  }
+
+  setTaxonThresholdFilterState() {
+    // Set the state of whether or not a taxon passes the custom threshold filters
+    // for each selected sample.
+    let {
+      sampleDetails,
+      allTaxonDetails,
+      allData,
+      taxonFilterState,
+    } = this.state;
+    Object.values(sampleDetails).forEach(sample => {
```

Comment:
Should we keep an aggregated state per sample and another per taxa to avoid going through all the options later?